File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <algorithm>
#include <vector>

using namespace std;

void maxDungeons(int k, vector<vector<int>>& dungeons, int cnt,
                vector<bool>& visited, vector<int>& result) {
    
    if(cnt > 0)
        result.push_back(cnt);
        
    for(int i = 0 ; i < dungeons.size() ; i++) {
        if(!visited[i] && k >= dungeons[i][0]) {
            visited[i] = true;
            maxDungeons(k - dungeons[i][1], dungeons, cnt + 1, visited, result);
            visited[i] = false;
        }
    }

}

int solution(int k, vector<vector<int>> dungeons) {
    vector<bool> visited(dungeons.size(), false);
    vector<int> result;
    maxDungeons(k, dungeons, 0, visited, result);
    sort(result.begin(), result.end());
    
    return result[result.size() - 1];
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        return getCnt(dungeons, new boolean[dungeons.length], k, new ArrayList<>());
    }
    
    int getCnt(int[][] dungeons, boolean[] visited, int k, List<int[]> res){
        if(res.size() == dungeons.length){
            int cnt = 0;
            for(int i = 0 ; i < res.size() ; i++){
                int[] cur = res.get(i);
                // 탐험이 불가능하면
                if(cur[0]>k){
                    break;
                }
                k -= cur[1];
                cnt++;
            }
            return cnt;
        }
        
        int ret = 0;
        
        // 순열
        for(int i = 0 ; i < dungeons.length ; i++){
            if(!visited[i]){
                visited[i] = true;
                res.add(new int[]{dungeons[i][0], dungeons[i][1]});
                ret = Math.max(ret, getCnt(dungeons, visited, k, res));
                res.remove(res.size() - 1);
                visited[i] = false;
            }
        }
        
        return ret;
    }
}